feat(core): Fix populating of node custom api call options

OlegIvaniv opened the pull request:
Github issue / Community forum post (link here to close automatically): https://community.n8n.io/t/shopify-generic-api-call/22375

OlegIvaniv committed:
feat(core): Fix populating of node custom api call options

OlegIvaniv committed:
lint fixes

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -19,6 +19,9 @@ import { CUSTOM_NODES_CATEGORY } from './Constants';
 import type { n8n } from './Interfaces';
 import { loadClassInIsolation } from './ClassLoader';
 
+const CUSTOM_API_CALL_NAME = 'Custom API Call';
+const CUSTOM_API_CALL_KEY = '__CUSTOM_API_CALL__';
```
There is a `Constants.ts` in this dir.

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -54,6 +57,58 @@ export abstract class DirectoryLoader {
+	protected supportsProxyAuth(description: INodeTypeDescription) {
+		if (!description.credentials) return false;
+
+		// const credentialTypes = CredentialTypes();
```
```suggestion

```

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -54,6 +57,58 @@ export abstract class DirectoryLoader {
+
+			if (credType.authenticate !== undefined) return true;
+
+			return this.isOAuth(credType);
```
Maybe we can inline this? Since it is not reused and is simple enough.

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -54,6 +57,58 @@ export abstract class DirectoryLoader {
+	 * Inject a `Custom API Call` option into `resource` and `operation`
+	 * parameters in a node that supports proxy auth.
+	 */
+	protected injectCustomApiCallOption(description: INodeTypeDescription) {
```
We can also keep this private until needed by child classes.

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -54,6 +57,58 @@ export abstract class DirectoryLoader {
+			return p;
+		});
+
+		return description;
```
Do we need to return the description if we are only relying on the mutation? Same for L92.

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -84,6 +139,7 @@ export abstract class DirectoryLoader {
 
 		if ('nodeVersions' in tempNode) {
 			for (const versionNode of Object.values(tempNode.nodeVersions)) {
+				this.injectCustomApiCallOption(versionNode.description);
```
Do we want to inject it to all versions or just to the latest? How do we know that a version has access to a cred that supports proxy auth?

OlegIvaniv replied:
That's a fair point as we don't have versioned credentials. I've changed it to only inject them for the latest version.

ivov reviewed (changes requested):
It'd be nice to add an e2e test to verify that the injected option is there, if not too time-consuming.

OlegIvaniv committed:
Adress PR comments

OlegIvaniv committed:
Add e2e test and only inject custom API options for latest version

OlegIvaniv commented:
@ivov Thanks for reviewing and for all the helpful suggestions. I've addressed them and added the e2e test to check for the Custom API option and hint. Can you have another look, please?

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -54,6 +54,48 @@ export abstract class DirectoryLoader {
+	 * Whether any of the node's credential types may be used to
+	 * make a request from a node other than itself.
+	 */
+	protected supportsProxyAuth(description: INodeTypeDescription) {
```
```suggestion
	private supportsProxyAuth(description: INodeTypeDescription) {
```

ivov commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -99,10 +141,11 @@ export abstract class DirectoryLoader {
-				? tmpNode.description.version.slice(-1)[0]
-				: tmpNode.description.version;
+
+			this.injectCustomApiCallOption(tempNode.description);
```
Agree with removing the injection from inside the loop so that not all node versions get the option, but if we want to inject only to the latest version, then the injection should still be inside the first branch and it's now missing it seems? 

As in, this second branch will only be reached if the node is not versioned, so I expect no versioned node is getting the custom option injected right now. But let me know if I'm misunderstanding.

OlegIvaniv replied:
You're right of course, I misunderstood how this method works. I've added`injectCustomApiCallOption` call also for `currentVersionNode`. Which now works as expected:
![Uploading CleanShot 2023-02-02 at 10.51.17@2x.png…
<img width="1342" alt="CleanShot 2023-02-02 at 10 51 36@2x" src="https://user-images.githubusercontent.com/12657221/216292032-ffb93312-ec54-494f-b3d7-c202112b6fd2.png">
]()

Thanks for catching this!

ivov commented on cypress/e2e/5-ndv.cy.ts:
```diff
@@ -87,4 +87,16 @@ describe('NDV', () => {
 		});
 	});
+
+	it('should show http node hint if node has custom api option', () => {
```
Love it, thank you.

ivov reviewed (commented):
Thanks for addressing everything!

OlegIvaniv committed:
Make sure to injectCustomApiCallOption for the latest version of node

ivov reviewed (approved):
🚀

netroy commented on packages/core/src/DirectoryLoader.ts:
```diff
@@ -54,6 +54,48 @@ export abstract class DirectoryLoader {
+		if (!description.credentials) return false;
+
+		return description.credentials.some(({ name }) => {
+			const credType = this.credentialTypes[name].type;
```
This check here assumes that all credentials are loaded before the nodes. This is breaking custom nodes. [Here](https://github.com/n8n-io/n8n/pull/5344) is the fix.
